File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.java
class Solution {
    public int[] sortedSquares(int[] nums) {

        // 배열 첫 번째가 0보다 크면 투포인터 알고리즘 사용할 필요 없다.
        if (check_all_bigger_0_or_smaller_0(nums)) {
            return nums;
        }

        int left = 0;
        int right = nums.length - 1;

        int[] result = new int[nums.length];
        int index = nums.length - 1;
        while (left <= right) {

            int leftValue = Math.abs(nums[left]);
            int rightValue = nums[right];

            if (leftValue <= rightValue) {
                result[index] = rightValue * rightValue;
                right--;
            } else {
                result[index] = leftValue * leftValue;
                left++;
            }
            index--;
        }

        return result;
    }

    private boolean check_all_bigger_0_or_smaller_0(int[] nums) {
        if (nums[0] >= 0) {
            for (int i = 0; i < nums.length; i++) {
                nums[i] = nums[i] * nums[i];
            }
            return true;
        } else if (nums[nums.length - 1] <= 0) {
            int[] copied = Arrays.copyOf(nums, nums.length);
            int i = 0, j = nums.length - 1;
            while (i < nums.length) {
                nums[i] = copied[j] * copied[j];
                i++;
                j--;
            }
            return true;
        }
        return false;
    }
}

File: 0977-squares-of-a-sorted-array/0977-squares-of-a-sorted-array.py
import collections
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        li = nums
        l = 0
        r = len(li)-1
        result = collections.deque([])
        
        while l <= r:
            if li[l] ** 2 < li[r] ** 2:
                result.appendleft(li[r] ** 2)
                r-=1
            else:
                result.appendleft(li[l] ** 2)
                l+=1
        return list(result)
